lib/rspec/fortify.rb

```diff
@@ -0,0 +1,255 @@
+
+        break unless should_retry?
+
+        ex.metadata[:retry_exceptions] << ex.exception if ex.exception
+
+        break if attempts >= retry_count
+
+        if exceptions_to_hard_fail.any? && ex.exception && exception_exists_in?(exceptions_to_hard_fail, ex.exception)
+          break
+        end
+
+        if exceptions_to_retry.any? && ex.exception && !exception_exists_in?(exceptions_to_retry, ex.exception)
+          break
+        end
+
+        if verbose_retry? && display_try_failure_messages? && retry_on_failure? && (attempts != retry_count)
+          exception_strings =
+            if ex.exception.is_a?(::RSpec::Core::MultipleExceptionError::InterfaceTag)
+              ex.exception.all_exceptions.map(&:to_s)
+            else
+              [ex.exception.to_s]
+            end
+
+          try_message = "\n#{ordinalize(attempts)} Try error in #{ex.location}:\n#{exception_strings.join "\n"}\n"
+          RSpec.configuration.reporter.message(try_message)
+        end
+
+        ex.example_group_instance.clear_lets if clear_lets
+
+        if RSpec.configuration.retry_callback
+          ex.ex_group_instance.instance_exec(ex, &RSpec.configuration.retry_callback)
+        end
+
+        sleep sleep_interval if sleep_interval.to_f.positive?
+      end
+    end
+
+    private
+
+    # borrowed from ActiveSupport::Inflector
+    def ordinalize(number)
+      if (11..13).include?(number.to_i % 100)
+        "#{number}th"
+      else
+        case number.to_i % 10
+        when 1; "#{number}st"
+        when 2; "#{number}nd"
+        when 3; "#{number}rd"
+        else    "#{number}th"
+        end
+      end
+    end
+
+    def exception_exists_in?(list, exception)
+      list.any? do |exception_klass|
+        exception.is_a?(exception_klass) || exception_klass === exception
+      end
+    end
+
+    def log_first_attempt?
+      cast_to_boolean(ENV.fetch('RSPEC_FORTIFY_LOG_FIRST_ATTEMPT', 'false'))
+    end
+
+    def retry_on_failure?
+      RSpec.configuration.retry_on_failure
+    end
+
+    def retry_on_success?
+      RSpec.configuration.retry_on_success && current_example_changed?
+    end
+
+    def current_example_changed?
+      # Added or modified specs can be passed in via this
+      # CHANGED_SPECS env var using some version of git diff like:
+      # git diff origin/main...HEAD --name-only --relative --diff-filter=AM | grep '_spec.rb$' | tr '\n' ',' | sed 's/,$//'
+      changed_specs = ENV.fetch('CHANGED_SPECS', nil)&.split(',') || []
+      changed_specs.include?(ex.file_path.sub(%r{^\./}, ''))
+    end
+
+    def current_attempt_failed?
+      !ex.exception.nil?
+    end
+
+    def current_attempt_succeeded?
+      ex.exception.nil?
+    end
+
+    def should_retry?
+      if retry_on_success?
+        current_attempt_succeeded?
+      elsif retry_on_failure?
+        current_attempt_failed?
+      end
+    end
+  end
+end
+
+def cast_to_boolean(value)
+  if %w(false f 0 no n).include?(value.to_s.downcase)
```

Comment:
What if `value` is `nil`? I'd expect it to cast to false, but this looks like it would cast to true.